File: Linked List/DCL.java
// Doubly Circular Linked List
class node
{
    public int data;
    public node next;
    public node prev;

    public node(int no)
    {
        this.data = no;
        this.next = null;
        this.prev = null;
    }
}

class DoublyCL
{
    private node first;
    private node last;
    private int iCount;

    public DoublyCL()
    {
        this.first = null;
        this.last = null;
        this.iCount = 0;
    }

    public void InsertFirst(int no)
    {
        node newn = null;

        newn = new node(no);

        if((this.first == null) && (this.last == null))
        {
            this.first = newn;
            this.last = newn;
        }
        else
        {
            newn.next = this.first;
            this.first.prev = newn;
            this.first = newn;
        }

        this.last.next = this.first;
        this.first.prev = this.last;

        this.iCount++;
    }

    public void InsertLast(int no)
    {
        node newn = null;

        newn = new node(no);

        if((this.first == null) && (this.last == null))
        {
            this.first = newn;
            this.last = newn;
        }
        else
        {
            newn.prev = this.last;
            this.last.next = newn;
            this.last = newn;
        }

        this.last.next = this.first;
        this.first.prev = this.last;

        this.iCount++;
    }

    public void DeleteFirst()
    {
        if((this.first == null) && (this.last == null))
        {
            return;
        }
        else if(this.first == this.last)
        {
            this.first = null;
            this.last = null;
        }
        else
        {
            this.first = this.first.next;
        }

        this.last.next = this.first;
        this.first.prev = this.last;

        this.iCount--;
    }

    public void DeleteLast()
    {
        if((this.first == null) && (this.last == null))
        {
            return;
        }
        else if(this.first == this.last)
        {
            this.first = null;
            this.last = null;
        }
        else
        {
            this.last = this.last.prev;
        }

        this.last.next = this.first;
        this.first.prev = this.last;

        this.iCount--;
    }

    public void Display()
    {
        node temp = null;

        temp = this.first;

        while(temp!=null)
        {
            System.out.print(" | "+temp.data+" | -> ");
            temp = temp.next;
        }
        System.out.println("null");
    }

    public int Count()
    {
        return this.iCount;
    }

    public void InsertAtPos(int no, int pos)
    {
        int iCnt = 0;

        node temp = null;
        node newn = null;

        if((pos<1) || (pos>iCount+1))
        {
            System.out.println("Invalid Position");
            return;
        }

        if(pos == 1)
        {
            InsertFirst(no);
        }
        else if (pos == iCount+1)
        {
            InsertLast(no);
        }
        else
        {
            newn = new node(no);

            temp = this.first;

            for(iCnt=1; iCnt<pos-1; iCnt++)
            {
                temp = temp.next;
            }
            newn.next = temp.next;
            temp.prev = newn;

            newn.prev = temp;
            temp.next = newn;
        }
    }

    public void DeleteAtPos(int pos)
    {
        int iCnt = 0;

        node temp = null;

        if((pos<1) || (pos>iCount))
        {
            System.out.println("Invalid Position");
            return;
        }

        if(pos == 1)
        {
            DeleteFirst();
        }
        else if (pos == iCount)
        {
            DeleteLast();
        }
        else
        {
            temp = this.first;

            for(iCnt=1; iCnt<pos-1; iCnt++)
            {
                temp = temp.next;
            }
            temp.next = temp.next.next;
        }        
    }
}

class DCL
{
    public static void main(String A[])
    {
        int iRet = 0;

        DoublyLL obj = null;
         
        obj = new DoublyLL();

        obj.InsertFirst(51);
        obj.InsertFirst(21);
        obj.InsertFirst(11);

        obj.Display();

        iRet = obj.Count();
        System.out.println("Number of nodes are: "+iRet);

        obj.InsertLast(101);
        obj.InsertLast(111);
        obj.InsertLast(121);

        obj.Display();

        iRet = obj.Count();
        System.out.println("Number of nodes are: "+iRet);

        obj.DeleteFirst();

        obj.Display();

        iRet = obj.Count();
        System.out.println("Number of nodes are: "+iRet);

        obj.DeleteLast();

        obj.Display();

        iRet = obj.Count();
        System.out.println("Number of nodes are: "+iRet);

        obj.InsertAtPos(105,4);

        obj.Display();

        iRet = obj.Count();
        System.out.println("Number of nodes are: "+iRet);

        obj.DeleteAtPos(4);

        obj.Display();

        iRet = obj.Count();
        System.out.println("Number of nodes are: "+iRet);
    }
}

File: Linked List/DCL.py
# Done
class Node:
    def __init__(self,value):
        self.data = value
        self.next = None
        self.prev = None

class DoublyCL:
    # Done 
    def __init__(self):
        self.first = None
        self.last = None
        self.iCount = 0

    # Done 
    def InsertFirst(self,no):
        newn = Node(no)

        # LL is empty
        if((self.first == None) and (self.last == None)):
            self.first = newn
            self.last = newn
        # LL has atleast one node
        else:
            newn.next = self.first
            self.first.prev = newn
            self.first = newn

        self.last.next = self.first
        self.first.prev = self.last

        self.iCount = self.iCount + 1    

    # Done 
    def InsertLast(self,no):
        newn = Node(no)

        # LL is empty
        if(self.first == None):
            self.first = newn
            self.last = newn
        # LL has atleast one node
        else:
            self.last.next = newn
            newn.prev = self.last
            self.last = newn 

        self.last.next = self.first 
        self.first.prev = self.last   

        self.iCount = self.iCount + 1           

    # Done 
    def InsertAtPos(self,no,pos):
        # Filter
        if(pos<1 or pos>(self.iCount+1)):
            print("Invalid Position")
            return

        if(pos == 1):
            self.InsertFirst(no)
            return
        elif(pos == (self.iCount+1)):
            self.InsertLast(no)
            return  
        else:
            newn = Node(no)
            temp = self.first

            for i in range(1,pos-1):
                temp = temp.next

            newn.next = temp.next
            temp.next.prev = newn

            temp.next = newn
            newn.prev = temp

            self.iCount = self.iCount + 1       

    # Done 
    def DeleteFirst(self):
        if((self.first == None) and (self.last == None)):
            return
        elif(self.first == self.last):
            self.first = None
            self.last = None
        else:
            self.first = self.first.next 
            del self.first.prev       

        self.last.next = self.first
        self.first.prev = self.last

        self.iCount = self.iCount - 1    

    # Done 
    def DeleteLast(self):
        # LL is empty
        if((self.first == None) and (self.last == None)):
            return

        # LL has one node
        elif(self.first == self.last):
            self.first = None
            self.last = None

        # LL contains more than one node
        else:
            self.last = self.last.prev
            del self.last.next

        self.last.next = self.first    

        self.iCount = self.iCount - 1 

    # Done 
    def DeleteAtPos(self,pos):
        # Filter
        if(pos<1 or pos>(self.iCount)):
            print("Invalid Position")
            return

        if(pos == 1):
            self.DeleteFirst()
            return
        elif(pos == (self.iCount)):
            self.DeleteLast()
            return 
        else:
            temp = self.first

            for i in range(1,pos-1):
                temp = temp.next

            temp.next = temp.next.next 
            del temp.next.prev
            temp.next.prev = temp

            self.iCount = self.iCount - 1  

    # Done 
    def Display(self):
        temp = self.first

        while(temp != self.last):
            print("| ",temp.data," |->",end = " ")
            temp = temp.next
            
        print("| ",temp.data," |-> None")    

    # Done 
    def Count(self):
        return self.iCount    

def main():
    sobj = DoublyCL()

    sobj.InsertFirst(101)
    sobj.InsertFirst(51)
    sobj.InsertFirst(21)
    sobj.InsertFirst(11)

    print("Elements of the Linked List are : ")
    sobj.Display()

    print("Number of elements in the Linked List are : ",sobj.Count())

    sobj.InsertLast(111)
    sobj.InsertLast(121)
    
    print("Elements of the Linked List are : ")
    sobj.Display()

    print("Number of elements in the Linked List are : ",sobj.Count())

    sobj.InsertAtPos(75,4)
    
    print("Elements of the Linked List are : ")
    sobj.Display()

    print("Number of elements in the Linked List are : ",sobj.Count())

    sobj.DeleteFirst()
    sobj.DeleteFirst()

    print("Elements of the Linked List are : ")
    sobj.Display()

    print("Number of elements in the Linked List are : ",sobj.Count())

    sobj.DeleteLast()

    print("Elements of the Linked List are : ")
    sobj.Display()

    print("Number of elements in the Linked List are : ",sobj.Count())

    sobj.DeleteAtPos(3)
    
    print("Elements of the Linked List are : ")
    sobj.Display()

    print("Number of elements in the Linked List are : ",sobj.Count())


if __name__ == "__main__":
    main()